[TensorPipe Agent] Bind device IP address

jiayisuse opened the pull request:
To resolve the issue: https://github.com/pytorch/pytorch/issues/36715

In tensorpipe rpc agent, we currently hardcoded localhost as pipes handshake IP address. This prevents us from setting up cross-host connections. As the first step, we start binding IP address for a given network device. For now it's defaulted to eth0. Will provide options to let user configure

Differential Revision: D21421094

lw reviewed (changes requested):
For the sake of consistency and predictability (on the users' side) I think we should use the same behavior as Gloo here. Which is the following:
https://github.com/facebookincubator/gloo/blob/b2ff1aa716dd37be5000dd3f13b214610c87d22f/gloo/transport/tcp/device.cc#L145-L158

So, basically, if the user specifies a device (for example using the `GLOO_SOCKET_IFNAME` env var, but there may be other ways), then it does indeed look up the address of that device. However, by default, it obtains the machine's hostname and resolves that to an IP address.

I think we should do the same. You could use the Gloo code I linked above as a reference.

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -65,6 +65,11 @@ class TensorPipeAgent : public RpcAgent {
+  // Retrieve IP address for a given network device for corss-hosts
+  // to set up tensorpipe connection. For now we default the device
+  // name eth0.
+  static std::string getDeviceIPAddress(const std::string& device = "eth0");
```
As I mentioned before, resolving IP address from default hostname will have localhost IP address. That doesn't work as expected from my experiment.

jiayisuse commented:
@lw As I mentioned before, resolving IP address from default hostname will have localhost IP address. That doesn't work as expected from my experiment.

jjlilley commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,6 +413,51 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
+
+  // Loop through all the results and pick up the first we can bind.
+  for (p = servinfo; p != NULL; p = p->ai_next) {
+    int fd = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
```
this fd needs to be closed() at some point, otherwise we leak descriptors

jjlilley commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,6 +413,51 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  rv = getaddrinfo(hostname, nullptr, &hints, &servinfo);
+  TORCH_INTERNAL_ASSERT(rv == 0, "Get address info error: ", gai_strerror(rv));
```
should we instead fallback if getaddrinfo() fails? Similar with gethostname()?

getaddrinfo() can be a bit flaky in prod, since it generates network traffic

lw replied:
Does it generate DNS traffic even when resolving its own hostname?

jiayisuse replied:
My guess is that it would go through DNS. Yeah we could fallback

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,6 +413,51 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
+  struct addrinfo hints, *servinfo, *p;
+  memset(&hints, 0, sizeof hints);
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
```
```suggestion
  hints.ai_socktype = SOCK_STREAM;
  hints.ai_protocol = IPPROTO_TCP;
```
Does it perhaps make sense to only ask for addresses that support TCP? I don't know if it ever makes a difference, but it shouldn't hurt either...

jiayisuse replied:
Just followed Gloo and it shouldn't hurt..

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,6 +413,51 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
+      struct sockaddr_in6* h = (struct sockaddr_in6*)p->ai_addr;
+      inet_ntop(AF_INET6, &h->sin6_addr, (char*)ipv6.data(), INET6_ADDRSTRLEN);
+      freeaddrinfo(servinfo);
+      return ipv6;
```
You cannot return early here or you will skip the `freeaddrinfo(servinfo);` call below and leak memory. Perhaps a RAII wrapper around it would be best here, so that we can also deal with exceptions properly.

jiayisuse replied:
On line 433 it's calling freeaddrinfo(servinfo)

lw replied:
Yes but if you *return* you're skipping all the rest of the function and thus you're also skipping the `freeaddrinfo`.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,6 +412,64 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
+    }
+    int bind_rv = bind(fd, p->ai_addr, p->ai_addrlen);
+    if (bind_rv == -1) {
+      continue;
```
You need to close the fd before you continue or it will leak.

jiayisuse committed:
[TensorPipe Agent] Bind default IP address (#37910)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37910

To resolve the issue: https://github.com/pytorch/pytorch/issues/36715

In tensorpipe rpc agent, we currently hardcoded localhost as pipes handshake IP address. This prevents us from setting up cross-host connections. As the first step, we start binding IP address for a given network device. For now it's defaulted to eth0. Will provide options to let user configure

Test Plan: CI

Reviewed By: lw

Differential Revision: D21421094

fbshipit-source-id: fa462fbdb441228a56c8a04a90d253cdb1f9564f
